storage/aptosdb/src/state_merkle_db.rs

```diff
@@ -657,6 +657,63 @@ impl StateMerkleDb {
 
         Ok(ret)
     }
+
+    fn get_rightmost_leaf_in_single_shard(
+        &self,
+        version: Version,
+        shard_id: u8,
+    ) -> Result<Option<(NodeKey, LeafNode)>> {
+        assert!(
+            shard_id < NUM_STATE_SHARDS as u8,
+            "Invalid shard_id: {}",
+            shard_id
+        );
+        let shard_db = self.state_merkle_db_shards[shard_id as usize].clone();
+        // The encoding of key and value in DB looks like:
+        //
+        // | <-------------- key --------------> | <- value -> |
+        // | version | num_nibbles | nibble_path |    node     |
+        //
+        // Here version is fixed. For each num_nibbles, there could be a range of nibble paths
+        // of the same length. If one of them is the rightmost leaf R, it must be at the end of this
+        // range. Otherwise let's assume the R is in the middle of the range, so we
+        // call the node at the end of this range X:
+        //   1. If X is leaf, then X.account_key() > R.account_key(), because the nibble path is a
+        //      prefix of the account key. So R is not the rightmost leaf.
+        //   2. If X is internal node, then X must be on the right side of R, so all its children's
+        //      account keys are larger than R.account_key(). So R is not the rightmost leaf.
+        //
+        // Given that num_nibbles ranges from 0 to ROOT_NIBBLE_HEIGHT, there are only
+        // ROOT_NIBBLE_HEIGHT+1 ranges, so we can just find the node at the end of each range and
+        // then pick the one with the largest account key.
+        let mut ret = None;
+
+        for num_nibbles in 1..=ROOT_NIBBLE_HEIGHT + 1 {
```

Comment:
Others wrote this originally but it is confusing: `num_nibbles` is actually `num_nibbles + 1`. Shall we fix it so that line 696 makes more sense?

Reply:
rename to `level` and add a comment

Reply:
I think it's more readable to do:
```
for num_nibbles in 0..=ROOT_NIBBLE_HEIGHT {
...
let seek_key = (version, num_nibbles + 1);
...
}
```